### ISSUE TITLE
[optimizers] Add KL-Shampoo (Adam-free stable Shampoo via KL-divergence minimization)

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#5633]
## Summary

Add **KL-Shampoo** — an Adam-free stable variant of the Shampoo optimizer — to `odyssey.training.optimizers`, alongside the landed Shampoo and SOAP.

KL-Shampoo (Lin et al., 2025, arXiv:2509.03378) recasts Shampoo's Kronecker-factor estimation as covariance estimation under **Kullback–Leibler (KL) divergence minimization** instead of the Frobenius-norm view underlying standard Shampoo. The KL objective respects the symmetric-positive-definite constraint on the preconditioner factors that the Frobenius norm ignores. The practical payoff: KL-Shampoo stabilizes **without** the in-eigenbasis Adam moments that SOAP relies on, eliminating Adam's first/second-moment memory overhead while matching or exceeding SOAP in pretraining.

## Update rule (paper Eq. 5, per step, for a 2-D weight `W (R×C)`)

```
S_A ← (1 − β) · S_A + (β / d_B) · G  S_B⁻¹ Gᵀ    # R×R, uses OLD S_B, d_B = C
S_B ← (1 − β) · S_B + (β / d_A) · Gᵀ S_A⁻¹ G     # C×C, uses OLD S_A, d_A = R
W   ← W − γ · S_A^{-1/2} G S_B^{-1/2}            # inverse-square-root preconditioned step
```

The distinguishing feature vs. standard Shampoo is the **coupled** update: each factor is updated with the gradient whitened by the *cross* factor's inverse (the KL stationarity condition ∂KL/∂S_A = 0, ∂KL/∂S_B = 0), not the raw Gram outer products `G Gᵀ` / `Gᵀ G`. Because each factor already absorbs the cross-factor whitening, the preconditioner uses the inverse **square** root per factor (not Shampoo's inverse fourth root).

## Scope

- Pure-functional module `src/odyssey/training/optimizers/kl_shampoo.mojo` (`kl_shampoo_step`, `kl_shampoo_step_simple`, `init_kl_shampoo_state`, `is_kl_shampoo_eligible`) mirroring the Shampoo/SOAP preconditioner family — caller owns and threads the two Kronecker factors.
- All optimizer state and every matrix root/inverse computed in **float64** (Odyssey `matmul` raises on mixed dtypes; the SOAP mixed-dtype lesson), with factor inverses/inverse-square-roots via `symmetric_eigh` (the same exact-eigenbasis approach SOAP uses).
- Unit test with a numpy parity reference (3-step state-evolution fixture) to the SOAP/Adan tolerance (1e-6).
- Registered in the primitive test runner and README primitive table; exported through `optimizers/__init__.mojo`.

## References

- Lin, W., Lowe, S. C., Dangel, F., Eschenhagen, R., Xu, Z., & Grosse, R. B. (2025). *Understanding and Improving Shampoo and SOAP via Kullback–Leibler Minimization.* arXiv:2509.03378.
- Depends conceptually on the landed Shampoo (Gupta et al. 2018 / Anil et al. 2020) and SOAP (Vyas et al. 2024) optimizers.

Tracking (consumer project): mvillmow/Random#75
